### ISSUE TITLE
if I want to change the neural networks, how can I do?

### ISSUE THREAD

**Issue by sumorday** [w86763777/pytorch-gan-collections#12]
HI, I want to test the  neural networks below:

![cnn](https://user-images.githubusercontent.com/49804057/168754384-50f6d37d-c8c1-481f-ae22-e1bf722d4eef.png)

And I tried to change the model files like below:

[class Generator(nn.Module):
    def __init__(self, z_dim, M):
        super(Generator, self).__init__()
        self.z_dim = z_dim
        self.main = nn.Sequential(
            nn.ConvTranspose2d(self.z_dim, 256, M, 1, 0, bias=False),  # 4, 4
            nn.BatchNorm2d(256),
            nn.ReLU(True),
            nn.ConvTranspose2d(256, 128, 4, stride=2, padding=1, bias=False),
            nn.BatchNorm2d(128),
            nn.ReLU(True),
            nn.ConvTranspose2d(128, 64, 4, stride=2, padding=1, bias=False),
            nn.BatchNorm2d(64),
            nn.ReLU(True),
            nn.ConvTranspose2d(64, 3, 4, stride=2, padding=1, bias=False),
            nn.Tanh()
        )

    def forward(self, z):
        return self.main(z.view(-1, self.z_dim, 1, 1))

class Discriminator(nn.Module):
    def __init__(self, M):
        super(Discriminator, self).__init__()

        self.main = nn.Sequential(
            # 32
            nn.Conv2d(3, 32, 5, 2, 2, bias=False),
            nn.LeakyReLU(0.2, inplace=True),
            # 16
            nn.Conv2d(32, 64, 5, 2, 2, bias=False),
            nn.LeakyReLU(0.2, inplace=True),
            nn.BatchNorm2d(64),
            # 8
            nn.Conv2d(64, 128, 5, 2, 2, bias=False),
            nn.LeakyReLU(0.2, inplace=True),
            nn.BatchNorm2d(128),
            # 4
            nn.Conv2d(128, 10, 5, 2, 2, bias=False), #128*2*2
            nn.ReLU(True),

        )

        self.linear = nn.Linear(M // 16 * M // 16 * 10, 1)

    def forward(self, x):
        x = self.main(x)
        x = torch.flatten(x, start_dim=1)
        x = self.linear(x)
        return x

class Generator32(Generator):
    def __init__(self, z_dim):
        super().__init__(z_dim, M=3)

class Generator48(Generator):
    def __init__(self, z_dim):
        super().__init__(z_dim, M=4)

class Discriminator32(Discriminator):
    def __init__(self):
        super().__init__(M=16)

class Discriminator48(Discriminator):
    def __init__(self):
        super().__init__(M=48)]

it seems I have to change the M and other parameters . How can I do? Thank you in advance

**Comment by sumorday** [comment]
[class Generator(nn.Module):
    def __init__(self, z_dim, M=4):
        super().__init__()
        self.M = M
        self.linear = nn.Linear(z_dim, M * M * 256)
        self.main = nn.Sequential(
            nn.BatchNorm2d(256),
            nn.ReLU(True),
            nn.ConvTranspose2d(256, 128, kernel_size=4, stride=2, padding=1),
            nn.BatchNorm2d(128),
            nn.ReLU(True),
            nn.ConvTranspose2d(128, 64, kernel_size=4, stride=2, padding=1),
            nn.BatchNorm2d(64),
            nn.ReLU(True),
            nn.Conv2d(64, 3, kernel_size=3, stride=1, padding=1),
            nn.Tanh())
        self.initialize()

    def initialize(self):
        for m in self.modules():
            if isinstance(m, (nn.Conv2d, nn.ConvTranspose2d, nn.Linear)):
                init.normal_(m.weight, std=0.02)
                init.zeros_(m.bias)

    def forward(self, z, *args, **kwargs):
        x = self.linear(z)
        x = x.view(x.size(0), -1, self.M, self.M)
        x = self.main(x)
        return x


class Discriminator(nn.Module):
    def __init__(self, M=32):
        super().__init__()
        self.M = M

        self.main = nn.Sequential(
            # M
            nn.Conv2d(3, 32, kernel_size=3, stride=1, padding=1),
            nn.LeakyReLU(0.1, inplace=True),
            nn.Conv2d(32, 64, kernel_size=4, stride=2, padding=1),
            nn.LeakyReLU(0.1, inplace=True),
            nn.BatchNorm2d(64),
            nn.Dropout2d(p=0.2),
            # M / 2
            nn.Conv2d(64, 128, kernel_size=3, stride=1, padding=1),
            nn.LeakyReLU(0.1, inplace=True),
            nn.BatchNorm2d(128),
            nn.Dropout2d(p=0.2),
            nn.Conv2d(128, 10, kernel_size=3, stride=1, padding=1),
            nn.ReLU(True))

        self.linear = nn.Linear(M // 2 * M // 2 * 10, 1) # here I am not sure
        self.initialize()

    def initialize(self):
        for m in self.modules():
            if isinstance(m, (nn.Conv2d, nn.Linear)):
                init.normal_(m.weight, std=0.02)
                init.zeros_(m.bias)
                spectral_norm(m)

    def forward(self, x, *args, **kwargs):
        x = self.main(x)
        x = torch.flatten(x, start_dim=1)
        x = self.linear(x)
        return x](url)

**Comment by w86763777** [comment]
If you want to use custom model, you have better to write a new one with same prototype of `__init__(self, z_dim)` for generator and `__init__(self)` for discriminator, and update `net_G_models` and `net_D_models` in training scripts to include your models.

**Comment by sumorday** [comment]
Hi I tried to write a new one with same prototype of __init__(self, z_dim) for discriminator but the main.py(dcgan) ALSO need to change something like 

    self.linear = nn.Linear(M // 16 * M // 16 * 10, 1)

I have to delete M And it may happen some errors like :
Expected 4-dimensional input for 4-dimensional weight [128, 256, 4, 4], but got 2-dimensional input of size [128, 128] instead

some codes dimensions are 128. so it may cause the errors from calculate matrix. I have no ideas how to do...

if I delete that M，it will cause so many mistakes...
